src/hermit.rs

```diff
@@ -17,39 +20,46 @@ impl From<io::Error> for Error {
 }
 
 pub struct Hermit<T: Config> {
-    config: T,
+    config: Rc<T>,
 }
 
 impl<T: Config> Hermit<T> {
     pub fn new(config: T) -> Hermit<T> {
-        Hermit { config: config }
+        Hermit { config: Rc::new(config) }
     }
 
-    pub fn current_shell(&self) -> Option<Shell> {
+    pub fn current_shell(&self) -> Option<Shell<T>> {
         self.config
             .current_shell_name()
-            .map(|shell_name| Shell::new(shell_name, self.config.root_path()))
+            .map(|shell_name| Shell::new(shell_name, self.config.clone()))
     }
 
     pub fn set_current_shell(&mut self, name: &str) -> Result<(), Error> {
         if self.config.does_shell_exist(name) {
-            self.config.set_current_shell_name(name).map_err(From::from)
+            match Rc::get_mut(&mut self.config) {
+                Some(config) => config.set_current_shell_name(name).map_err(From::from),
+                None => {
+                    unreachable!(message::error("attempted to modify config while it was being \
```

Comment:
yeah, AFAIK it's basically a synonym for `panic!`, but it's more self-documenting about the expectation that it's code that shouldn't ever be reached.